### PR TITLE
handle MaterializeResult output typing with checks

### PR DESCRIPTION
**Pull request by alangenfeld**
update expectations around return typing to handle `MaterializeResult`

## How I Tested These Changes

updated tests

### CODE REVIEW

**Comment by alangenfeld** [comment]
Current dependencies on/for this PR:
* master
  * **PR #17039** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/17039" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/dagster-io/dagster/17039?utm_source=stack-comment).

File: python_modules/dagster/dagster/_core/definitions/op_definition.py [hunk 1]
```diff
@@ -481,35 +481,41 @@ def _resolve_output_defs_from_outs(
         only_out = outs[name]
         return [only_out.to_definition(annotation, name, description, default_code_version)]
 
-    output_defs: List[OutputDefinition] = []
+    # If no type hint or a result object type, use empty annotation
```
Comment (schrockn) [inline python_modules/dagster/dagster/_core/definitions/op_definition.py:484]: does this need to handle `AssetCheckResult`?  What about a future `ObservationResult`?

**Review by schrockn — CHANGES_REQUESTED** [review]
Bouncing to your q re: question about generalized result types

File: python_modules/dagster/dagster/_core/definitions/op_definition.py (hunk 1 above)
Comment (alangenfeld) [reply python_modules/dagster/dagster/_core/definitions/op_definition.py:484]: > does this need to handle AssetCheckResult?

I dont think so, singular `AssetCheckResult` should only be from `@asset_check` . i could test it 

> What about a future ObservationResult?

yes i assume so

Comment (schrockn) [reply python_modules/dagster/dagster/_core/definitions/op_definition.py:484]: can we make this a helper function or do something to make it clear where a future result type should live?

**Review by schrockn — APPROVED** [review]
One last comment otherwise lg!

**Pull request merged**